src/Functions/PartitionByHyperplanes.cpp

```diff
@@ -0,0 +1,338 @@
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        FunctionArgumentDescriptors mandatory_args{
+            {"vectors", &isArray<IDataType>, nullptr, "Array"},
+            {"normals", &isArray<IDataType>, nullptr, "Array"},
+        };
+
+        FunctionArgumentDescriptors optional_args{
+            {"offsets", &isArray<IDataType>, nullptr, "Array"},
+        };
+
+        validateFunctionArgumentTypes(*this, arguments, mandatory_args, optional_args);
+
+        return std::make_shared<DataTypeString>();
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t input_rows_count) const override
+    {
+        const ColumnArray * vectors = typeid_cast<const ColumnArray *>(arguments[0].column.get());
+        if (!vectors)
+            throw Exception(ErrorCodes::ILLEGAL_COLUMN, "First argument of function {} must be Array", getName());
+        const auto & nested_vectors_data = typeid_cast<const ColumnFloat32 &>(vectors->getData());
+
+        const ColumnArray * normals = typeid_cast<const ColumnArray *>(arguments[1].column.get());
+        if (!normals)
+            throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Second argument of function {} must be Array", getName());
+        const auto & nested_normals_data = typeid_cast<const ColumnFloat32 &>(normals->getData());
+
+        const size_t vector_size = vectors->getOffsets().front();
+
+        auto const_offsets = ColumnConst::create(ColumnFloat32::create(1, 0), input_rows_count * vector_size);
+        const IColumn * nested_offsets_data = const_offsets.get();
+        if (arguments.size() >= 3)
+        {
+            const ColumnArray * offsets = typeid_cast<const ColumnArray *>(arguments[2].column.get());
+            if (!offsets)
+                throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Third argument of function {} must be Array", getName());
+            nested_offsets_data = &offsets->getData();
+        }
+
+        if (input_rows_count == 0)
+            return result_type->createColumn();
+
+        checkDimension(vectors->getOffsets(), vector_size);
+        checkDimension(normals->getOffsets(), vector_size);
+
+        auto col_res = ColumnFloat32::create(input_rows_count * input_rows_count);
+        auto col_res_offsets = ColumnArray::ColumnOffsets::create();
+        col_res_offsets->reserve(input_rows_count);
+        for (size_t i = 0; i < input_rows_count; ++i)
+        {
+            col_res_offsets->insertValue(input_rows_count * i);
+        }
+
+        executeInternal(
+            nested_vectors_data,
+            nested_normals_data,
+            nested_offsets_data,
+            input_rows_count,
+            vector_size,
+            *col_res);
+
+        return ColumnArray::create(std::move(col_res), std::move(col_res_offsets));
+    }
+};
+
+}
+
+REGISTER_FUNCTION(partitionByHyperplanes)
```

Comment:
It's also good to add test for the function to https://github.com/ClickHouse/ClickHouse/blob/master/tests/queries/0_stateless/ to check implementation and how function works on valid and non-valid inputs

Reply:
@vdimir I added a few tests

Reply:
Adding more cases will be good. Some corner cases - empty array passed, dimensions are not matched, nested or flattened arrays and so on. Also we can check for some small and big dimentions, e.g. 1,2,3 and 1024, 10000

Reply:
Also, I'm not sure which target is run in ci also worth checking if all branches are covered (not sure if it's possible to have avx in ci, though)

Reply:
@vdimir I added more tests, and the code itself changed a bit.
Could you check it please?

Reply:
Also, is there any compile-time flag for checking OS? Something like `#ifdef linux`?
AMX version will compile only for linux

Reply:
> Also, is there any compile-time flag for checking OS?

Yes https://github.com/ClickHouse/ClickHouse/blob/2f1fb6c5d83d08e904dae1e1e8d4bd2297d2a0f3/src/Daemon/BaseDaemon.cpp#L16